Support modification of strict dependencies in component metadata rules

jjohannes opened the pull request:
https://github.com/gradle/gradle/issues/3447

melix commented on subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/DependencyMetadataRulesIntegrationTest.groovy:
```diff
@@ -213,7 +213,7 @@ abstract class DependencyMetadataRulesIntegrationTest extends AbstractHttpDepend
                             withDependencies {
                                 it.each {
-                                    it.version = '1.0'
+                                    it.version { prefer = '1.0' }
```
I am surprised this works. `prefer '1.0'` should pass, but `setPrefer('1.0')`?

jjohannes replied:
Good catch. This must be groovy magic. The _private_ field in `DefaultMutableVersionConstraint` is called `prefer`.

melix replied:
Ah yes that must be it. I guess we need to change the test to use `prefer '1.0'` for idiomatic Gradle.

jjohannes replied:
I thought I had changed it... I think I had two occurrences. Should be fixed now.

melix commented on subprojects/core-api/src/main/java/org/gradle/api/artifacts/DependencyMetadata.java:
```diff
@@ -43,12 +44,13 @@
      * which usually expresses what API level of the module you are compatible with.
      */
-    String getVersion();
+    VersionConstraint getVersion();
```
:x: Let's call this `getVersionConstraint`

jjohannes replied:
I called it `version` because it's called `version {}` for modifying and in the metadata. But.. it's different to `Dependency.getVersion()`. So I guess `getVersionConstraint ()` is what we want. I changed it.

jjohannes commented:
@melix I fixed the two things you commented on. The compatibility report is reporting the changes. But it is not failing, because for incubating API we do not fail (only report). See:
https://builds.gradle.org/viewLog.html?buildId=9358966&buildTypeId=Gradle_Check_SanityCheck&tab=report_project951_API_Compatibility_Report

I also fixed the Gradle build, which already uses the new metadata rule APIs.
